agent/php_error.c

```diff
@@ -561,12 +561,39 @@ static int nr_php_should_record_error(int type, const char* format TSRMLS_DC) {
     return 0;
   }
 
+  /*
+   * For a the following error types:
+   * E_WARNING || E_CORE_WARNING || E_COMPILE_WARNING || E_USER_WARNING
+   * PHP triggers an exception if EH_THROW is toggled on and then immediately
+   * returns after throwing the exception. PHP 7 additionally triggers an exception 
+   * for E_RECOVERABLE_ERROR.
+   * See for more info:
+   * https://github.com/php/php-src/blob/master/main/main.c In that case, we
+   * should not handle it, but we should exit and let the exception handler
+   * deal with it; otherwise, we could record an error even if an exception is
+   * caught.
+   */
+  if (EG(error_handling) == EH_THROW) {
+    if ((E_WARNING == type) || (E_CORE_WARNING == type)
+        || (E_COMPILE_WARNING == type) || (E_USER_WARNING == type)) {
+      return 0;
+    }
+#if ZEND_MODULE_API_NO < ZEND_8_0_X_API_NO
+/*
+ * Cover one more error condition that was falling through to default
+ * https://github.com/php/php-src/blob/PHP-7.4.17/main/main.c#L1248
+ */
+    if (E_RECOVERABLE_ERROR == type) {
+      return 0;
+    }
+#endif
+  }
```

Comment:
This will work, but this code's performance can be improved upon. Because of how Zend defines error types, they can be used as bits in a bit mask - see [here](https://github.com/php/php-src/blob/PHP-8.0/Zend/zend_errors.h#L48). Therefore I suggest taking advantage of that, and make the error type comparison once not 4 or 5 times. Defining these two macros:
```
#if ZEND_MODULE_API_NO < ZEND_8_0_X_API_NO
#define E_ERRORS_THROWING_EXCEPTIONS (E_WARNING | E_CORE_WARNING | E_COMPILE_WARNING | E_USER_WARNING | E_RECOVERABLE_ERROR)
#else
#define E_ERRORS_THROWING_EXCEPTIONS (E_WARNING | E_CORE_WARNING | E_COMPILE_WARNING | E_USER_WARNING)
#endif

#define E_THROWS_EXCEPTION(e) ((E_ERRORS_THROWING_EXCEPTIONS & e) == e)
```
simplifies the code to:
```suggestion
  if (EG(error_handling) == EH_THROW && E_THROWS_EXCEPTION(type)) {
    // let the exception handler deal with this error
    return 0;
  }
```
which, not only is more performant, but also more readable.